allennlp/data/samplers/bucket_batch_sampler.py

```diff
@@ -33,8 +33,12 @@ class BucketBatchSampler(BatchSampler):
     data_source: `data.Dataset`, required,
         The pytorch `Dataset` of allennlp Instances to bucket.
 
-    batch_size : `int`, required.
-        The size of each batch of instances yielded when calling the dataloader.
+    batch_size : `int`, optional (default = None)
+        The size of each batch of instances yielded when calling the dataloader. Cannot be used with
+        `max_tokens`.
+
+    max_tokens : `int`, optional (default = None)
+        The maximum number of tokens to include in a batch. Cannot be used with `batch_size`.
```

Comment:
You added this because huggingface uses this?

Reply:
After having read the whole thing, I don't like that this is part of `BucketIterator`. I think it should be its own type of iterator.

Of course, for pure research code, adding this functionality like this would be great. But this is not that case. This code actually has many users, and most of them never care about the `max_tokens` scenario, so I want to keep the API clean and easy for those guys.

Reply:
I created a subclass of `BucketBatchSampler` called `MaxTokensBatchSampler` and reverted `BucketBatchSampler`. I kept the changes for the argsort `_argsort_by_padding`, since this only returns the lengths alongside the indices.